### ISSUE TITLE
kube-router not working as expected

### ISSUE THREAD

**Issue by felipejfc** [cloudnativelabs/kube-router#93]
I have a scenario where I have 19 nodes, running like 900 pods + 900 services, a service per pod, I'm tying kube-router and I've realised that some of the services (type nodeport) are not working, I can telnet to pods directly but not through the service:

```
Name:			tan-07cdbae3
Namespace:		tan
Labels:			name=tan-07cdbae3
Annotations:		<none>
Selector:		app=tan-07cdbae3
Type:			NodePort
IP:			100.64.100.200
Port:			pom-port	30000/TCP
NodePort:		pom-port	32678/TCP
Endpoints:		100.96.11.53:30000
Port:			room-port	10000/UDP
NodePort:		room-port	31135/UDP
Endpoints:		100.96.11.53:10000
Session Affinity:	None
Events:			<none>
```

```
$ telnet tan-07cdbae3.tan 30000
Trying 100.64.100.200...
^C

$ telnet 100.96.11.53 30000
Trying 100.96.11.53...
Connected to 100.96.11.53.
Escape character is '^]'.
```

As a side-effect I cant access the service through the nodeports.

This is not the only case that is not working as expected, in another cases, I can access a svc type NodePort from some nodes but in others the port is closed...

I really would like to use kube-router in production as it fits my need for performance and low overhead, but unfortunately, with this kind of instability it's impossible.

Any hints @murali-reddy ?

Thanks

**Comment by felipejfc** [comment]
maybe related to #57?

**Comment by murali-reddy** [comment]
@felipejfc sorry about the issues you are running into. So far me and @bzub have been running Kubernetes e2e tests to catch the nagging bugs. Honestly we did not run the tests at the scale of yours. We would definitely attempt to run tests at scale ASAP to address any instability. We have been focussing on testing/CD to bring the project production grade reliability and stability before moving Beta release.

Coming to this bug, do you happen to see any errors in the kube-router logs, related service not getting created? I will attempt to reporduce the bug and share my findings.

**Comment by murali-reddy** [comment]
@felipejfc quick update

so wrote [small script](https://gist.github.com/murali-reddy/dfc1fa7d00329029b711a2f2eddd8f37#file-ipvs-py) to test deploying large number of services. I am able to repro the problem

Network service controller thread in Kube-router dies due to fatal failure in nlgo library used by Kube-router. We have an issue #78 reported for this. 

We recently updated dependent libraries, which forced us to update nlgo to latest version. 

I am able to repro the issue at just 10 services. 

So basically network service controller stops functioning due to nlgo's error `genl hub loop exit`

I will investigate further and run some scale tests and get back to you.

**Comment by felipejfc** [comment]
Nice @murali-reddy, if we can repro it then we can fix it :) 
I'll be waiting for updates on this.

Thanks!!

**Comment by bzub** [comment]
@felipejfc If possible, please test the `cloudnativelabs/kube-router-git:PR101` container image. It should fix the issue. If you see log formatting bugs it is a known cosmetic issue at this time. Thanks!

**Comment by murali-reddy** [comment]
@bzub thats wonderful. 
I tested image with 1600 NodePort services deployed on 11 nodes cluster on AWS. They ran perfectly. Pleasse see [PR](https://github.com/cloudnativelabs/kube-router/issues/101) for test results.

**Comment by jsravn** [comment]
I'm curious since I'm running into the same problem with libipvs - did you figure out the actual issue in libipvs?

**Comment by murali-reddy** [comment]
@jsravn We hit this issue long time ago when kube-router was using https://github.com/mqliang/libipvs for IPVS, then we switched to docker/libnetwork. We never ran into this issue with docker/libnetwork.

**Comment by g0194776** [comment]
@murali-reddy 

Can you give me a version of kube-router that had fixed this problem? (Including changed to libnetwork)
